File: src/shady/builtin/embed.c
#include <stdio.h>
#include <string.h>
#include <stddef.h>

typedef enum {
    STRING, BIN
} Mode;

int main(int argc, char** argv) {
    if (argc != 5) {
        goto usage_fail;
    }

    Mode m;
    if (strcmp(argv[1], "string") == 0)
        m = STRING;
    else if (strcmp(argv[1], "bin") == 0)
        m = BIN;
    else
        goto usage_fail;

    const char* object_name = argv[2];

    FILE* src = fopen(argv[3], "rb");
    FILE* dst = fopen(argv[4], "wb");
    if (!src || !dst)
        goto io_fail;

    fseek(src, 0, SEEK_END);
    size_t size = ftell(src);
    fseek(src, 0, 0);

    printf("source file is %zd bytes long\n", size);

    size_t pos = 0;
    char buffer[16384];
    size_t read;

    fprintf(dst, "const char %s[] = ", object_name);
    switch (m) {
        case STRING: {
            fprintf(dst, "\"");
            while (pos < size) {
                read = fread(buffer, sizeof(char), sizeof(buffer) / sizeof(char), src);
                for (size_t i = 0; i < read; i++) {
                    char c = buffer[i];
                    switch (c) {
                        case '"':
                            fprintf(dst, "\\\"");
                            break;
                        case '\n':
                            fprintf(dst, "\\n\"\n\"");
                            break;
                        default:
                            fprintf(dst, "%c", c);
                            break;
                    }
                    pos++;
                }
                if (read == 0)
                    break;
            }
            fprintf(dst, "\"");
            break;
        }
        case BIN: {
            fprintf(dst, "{");
            while (pos < size) {
                read = fread(buffer, sizeof(char), sizeof(buffer) / sizeof(char), src);
                for (size_t i = 0; i < read; i++) {
                    char c = buffer[i];
                    fprintf(dst, "%d, ", c);
                    pos++;
                }
                if (read == 0)
                    break;
            }
            fprintf(dst, "}");
            break;
        }
    }
    fprintf(dst, ";\n");
    fclose(dst);

    return 0;

    io_fail:
    printf("i/o failure\n");
    return 2;

    usage_fail:
    printf("Usage: embedder [string | bin] name src_file dst_file\n");
    return 1;
}

File: src/common/embed.py
import io
import sys

argv = sys.argv
argc = len(argv)

def usage_fail():
    print("Usage: embedder.py [string | bin] name src_file dst_file")
    exit(1)

if argc != 5:
    usage_fail()

global mode
if argv[1] == "string":
    mode = argv[1]
elif argv[1] == "bin":
    mode = argv[1]
else:
    usage_fail()

object_name = argv[2]

src = open(argv[3], "rb")
dst = open(argv[4], "wb")

src.seek(0, io.SEEK_END)
size = src.tell()
src.seek(0, 0)

print(f"source file is {size} bytes long")

buffer = bytearray(src.read())

dst.write(f"const char {object_name}[] = ".encode())
if mode == "string":
    dst.write(f"\"".encode())
    for pos in range(0, size):
        c = buffer[pos]
        if c == "\"".encode()[0]:
            dst.write("\\\"".encode())
        elif c == "\n".encode()[0]:
            dst.write("\\n\"\n\"".encode())
        else:
            dst.write(bytes([c]))
        #dst.write(bytes(c))
        #dst.write(f"{c}\\".encode())
    dst.write(f"\"".encode())
elif mode == "bin":
    dst.write("{".encode())
    for pos in range(0, size):
        c = int(buffer[pos])
        dst.write(f"{c}, ".encode())
    dst.write("}".encode())

dst.write(f";\n".encode())
dst.close()